### PR TITLE
:bug: Avoid infinite loop nesting copies inside components

### PR DESCRIPTION
**Pull request by hirunatan**
Fixes https://tree.taiga.io/project/penpot/issue/5314

### CODE REVIEW

File: common/src/app/common/pages/changes.cljc [hunk 1]
```diff
@@ -385,15 +385,27 @@
 
 (defmethod process-change :mov-objects
   [data {:keys [parent-id shapes index page-id component-id ignore-touched after-shape]}]
-  (letfn [(calculate-invalid-targets [objects shape-id]
+  (letfn [(nested-components? [objects shape-id]
+             (let [children          (cph/get-children-with-self objects shape-id)
+                   child-components  (into #{} (keep :component-id children))
```
Comment (niwinz) [inline common/src/app/common/pages/changes.cljc]: `(into #{} (keep :component-id children))` can be replaced with: `(into #{} (keep :component-id) children)`

The `(keep :component-id)` can be also defined at namespace level for reuse like:

`(def xf-keep-component-id (keep :component-id))`

And then re-used:

`(into #{} xf-keep-component-id children)`

File: common/src/app/common/pages/changes.cljc [hunk 2]
```diff
@@ -385,15 +385,27 @@
...
+                   parents           (cph/get-parents-with-self objects parent-id)
+                   parent-components (into #{} (->> parents
+                                                    (filter :main-instance?)
+                                                    (map :component-id)))]
```
Comment (niwinz) [inline common/src/app/common/pages/changes.cljc]: here you also can replace

```
(into #{} (->> parents
               (filter :main-instance?)
               (map :component-id)))
```
with something like this

```
(let [xform (comp (filter :main-instance?) 
                  xf-keep-component-id)]
  (into #{} xform parents))
```

Remember, that transducers is a form of reusable and composable way to define transformations independently of the underlying datastructure.

File: common/src/app/common/pages/changes.cljc (hunk 1 above)
Comment (hirunatan) [reply common/src/app/common/pages/changes.cljc]: done

File: common/src/app/common/pages/changes.cljc (hunk 2 above)
Comment (hirunatan) [reply common/src/app/common/pages/changes.cljc]: transducers are my nemesis...

perhaps some day i will always remember to use them

**Pull request merged**